soundcard/pulseaudio.py

```diff
@@ -367,11 +354,54 @@ def record(self, numframes):
                     chunk = numpy.zeros(nbytes_ptr[0]//4, dtype='float32')
                 if nbytes_ptr[0] > 0:
                     self._pulse._pa_stream_drop(self.stream)
-                    captured_data.append(chunk)
-                    captured_frames += len(chunk)/self.channels
+                    return chunk
             else:
                 time.sleep(0.001)
-        return numpy.reshape(numpy.concatenate(captured_data), [-1, self.channels])
+
+    def record(self, numframes=None):
+        """Record some audio data.
+
+        The data will be returned as a `frames × channels` float32
+        numpy array.
+
+        This function will wait until `numframes` frames have been
+        recorded. However, the audio backend holds the final authority
+        over how much audio data can be read at a time.
+        The buffering is made inside the function so that exactly `numframes`
+        samples are returned at every call.
+        For real-time settings, set the blocksize attribute of the recorder
+        object to be smaller than the requested number of frames.
+
+        """
+        if numframes:
+            captured_data = [self._pending_chunk]
+            captured_frames = self._pending_chunk.shape[0] / self.channels
+            if captured_frames >= numframes:
+                keep , self._pending_chunk = numpy.split(self._pending_chunk,
+                [int(numframes * self.channels)])
+                return numpy.reshape(keep, [-1, self.channels])
+            else:
+                while captured_frames < numframes:
+                    chunk = self._record_chunk()
+                    captured_data.append(chunk)
+                    captured_frames += len(chunk)/self.channels
+                to_split = int(len(chunk) - (captured_frames - numframes) * self.channels)
+                _ = captured_data.pop()
+                keep, self._pending_chunk = numpy.split(chunk, [to_split])
+                captured_data.append(keep)
+                return numpy.reshape(numpy.concatenate(captured_data), [-1, self.channels])
+        else:
+            return numpy.reshape(self._record_chunk(), [-1, self.channels])
```

Comment:
I would move this branch up to the `if` (and change the `if` to `if numframes is None`). It is kind of confusing to find a short `else` after this much code.

Reply:
Also, this should include the previous `self._pending_chunk`, if any. (Or clear it, and mention that in the docstring)